text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,1417 @@
+              status_message: string | string_dictionary
+```
+
+## Implementation Recommendations
+
+### Protocol Extension and Fallback Mechanism
+
+The support of this new protocol can only be progressive, so implementers are advised to follow the following
+implementation recommendations in phase 1:
+
+* OTLP Receiver: Listen on a single TCP port for both OTLP and OTLP Arrow. The goal is to make the support of this
+  protocol extension
+  transparent and automatic. This can be achieved by adding the `ArrowStreamService` to the same gRPC listener. A
+  configuration
+  parameter will be added to the OTLP receiver to disable this default behavior to support specific uses.
+* OTLP Exporter: By default the OTLP exporter should initiate a connection to the `ArrowStreamService` endpoint of the target
+  receiver. If this connection fails because the `ArrowStreamService` is not implemented by the target, the exporter
+  must automatically fall back on the behavior of the classic OTLP protocol. A configuration parameter could be added to
+  disable this default behavior.
+
+The implementation of these two rules should allow a seamless and
+adaptive integration of OTLP Arrow into the current ecosystem
+generally.
+
+For the prototype specifically, which is a fork of the OpenTelemetry
+collector codebase, we have derived the OTLP/gRPC-Arrow exporter and
+receiver as set of changes directly to the `receiver/otlpreceiver` and
+`exporter/otlpexporter` components, with new `internal/arrow` packages
+in both.  With every collector release we merge the OTLP-Arrow changes
+with the mainline components to maintain this promise of
+compatibility.
+
+OTLP-Arrow supports conveying the gRPC metadta (i.e., http2 headers) using a dedicated `bytes` field.  Metadata is
+encoded using [hpack](https://datatracker.ietf.org/doc/rfc7541/) like a typical unary gRPC request.
+
+Specifically:
+
+#### OTLP/gRPC Receiver
+
+When Arrow is enabled, the OTLP receiver listens for both the standard unary gRPC service OTLP and OTLP-Arrow stream
+services.  Each stream uses an instance of the OTel-Arrow-Adapter's
+[Consumer](https://pkg.go.dev/github.com/f5/otel-arrow-adapter@v0.0.0-20230112224802-dafb6df21c97/pkg/otel/arrow_record#Consumer). Sets
+`client.Metadata` in the Context.
+
+#### OTLP/gRPC Exporter
+
+When Arrow is enabled, the OTLP exporter starts a fixed number of streams and repeatedly sends one `plog.Logs`,
+`ptrace.Traces`, or `pmetric.Metrics` item per stream request.  The `exporterhelper` callback first tries to get an
+available stream, blocking when none are available (or until the connection is downgraded), and then falls back to the
+standard unary gRPC path.  The stream-sending mechanism internally handles retries when failures are caused by streams
+restarting, while honoring the caller's context deadline, to avoid delays introduced by allowing these retries to go
```

Comment:
>while honoring the caller's context deadline

Please clarify what this means. Honoring, in the sense of ailing when exceeding?

Reply:
The diagram [here](https://github.com/open-telemetry/experimental-arrow-collector/blob/main/exporter/otlpexporter/arrow-design.png) tries to illustrate how each call into the standard or Arrow OTLP exporter goes through `exporterhelper` and the timeout and retry semantics that it uses should be "respected" (what I meant by honored) in the sense that the caller enters a `select` statement that includes a case for the caller's `ctx.Done`. 

(We probably do not need a RetryInfo as shown in the draft protocol, if we use whatever other mechanism the `exporterhelper` is using here.)